docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,119 @@
+it does not seem possible, in general, to verify that the chunk corresponds
+to the chunk number, even if it can be verified that the chunk is a valid set of
+nodes in the tree. A malicious peer could provide a chunk that contains valid
+nodes from the tree but does not actually match the alleged chunk number. This
+could make it difficult for nodes to figure out which parts of the tree they
+have and which parts they still need. Hence, to go with this chunking approach
+to state sync, information about how state is chunked would need to be hashed
+into the state itself, or into the block header, so that chunk numbers could be
+directly verified.
+
+An alternative design was suggested by Jae Kwon in
+[#3639](https://github.com/tendermint/tendermint/issues/3639) where chunking
+happens lazily and in a dynamic way: nodes request key ranges from their peers,
+and peers respond with some subset of the
+requested range and with notes on how to request the rest in parallel from other
+peers. Unlike chunk numbers, keys can be verified directly. And if some keys in the
+range are ommitted, proofs for the range will fail to verify.
+This way a node can start by requesting the entire tree from one peer,
+and that peer can respond with say the first few keys, and the ranges to request
+from other peers.
+
+Finally, there has been significant work over the last few months rearchitecting
+the blockchain reactor to make it more easily testable and verifiable. This has
+culminated in a most recent design,
+[#3575](https://github.com/tendermint/tendermint/pull/3753),
+where the reactor is split into three indpendent routines: one for IO with
+peers; one for reading, writting, and verifying blocks; and one for scheduling
+requests to peers. Since the high level goals of state sync are quite similar
+to block sync, the same architecture and separation of concerns should apply.
+
+### Considerations
+
+- Should we also support a warp-sync version? It would be faster
+  and maybe easier for tree implementers, but less secure (relies on
+  honest-majority of peers, rather than light-client)
+- Should we include info about state sync into the state itself so we can use
+  numbered chunks instead of key ranges? It would make the protocol simpler and
+  faster, but at the cost of a large precompute step and some more coupling with
+  the state or block header.
+
+### Next Steps
+
+To move forward with a light-client secure state sync based on key-ranges, here
+are some next steps:
+
+- The state sync reactor should follow the same architecture as the
+  new blockchain reactor ala [#3753](https://github.com/tendermint/tendermint/pull/3753).
+- The proposal in issue #3639 should be fleshed out and written up more formally as an ADR (ie.
+  in this one)
+- Need to address the initial light client sync (this will depend on a working
+  light client!)
+- Need to address assembling the Tendermint `State` object
```

Comment:
>  - Need to address assembling the Tendermint `State` object

From my understanding this means trying to figure out how to take TM state chunks and  resurrect the TM `State` object?

If so can we change to:
 ```suggestion
- Need to address assembling the Tendermint `State` object from `StateChunk` objects
```

Reply:
Not necessarily. Note this is a much smaller object than what we have to sync from the application. Also it's possible that the light client sync will give us everything we need here (ie. the validator set and consensus params). So we might not actually have to sync much else ...